src/repo.rs

```diff
@@ -490,43 +428,175 @@ impl<T: NetworkAdapter + 'static> Repo<T> {
-                self.process_outgoing_network_messages(&collection_id);
-            }
 
             loop {
                 select! {
                     recv(self.collection_receiver) -> collection_event => {
                         if let Ok((collection_id, event)) = collection_event {
+                            tracing::trace!(?collection_id, ?event, "received collection event");
                             self.handle_collection_event(&collection_id, event);
-                            self.process_outgoing_network_messages(&collection_id);
+                            self.sync_document_collection(&collection_id);
+                            self.fill_outboxes();
                         } else {
                             // The repo shuts down
                             // once all handles and collections drop.
                             break;
                         }
                     },
-                    recv(self.wake_receiver) -> event => {
-                        match event {
-                            Err(_) => panic!("Wake senders dropped"),
-                            Ok(WakeSignal::Stream(collection_id)) => {
-                                self.collect_network_events(&collection_id);
-                                self.sync_document_collection(&collection_id);
-                                self.process_outgoing_network_messages(&collection_id);
+                    recv(self.control_receiver) -> control_event => {
+                        match control_event {
+                            Ok(ControlEvent::CreateCollection{observer, reply}) => {
+                                let collection = self.new_collection(observer);
+                                reply.send(collection).unwrap();
                             },
-                            Ok(WakeSignal::Sink(collection_id)) => {
-                                self.process_outgoing_network_messages(&collection_id);
+                            Ok(ControlEvent::Shutdown) => {
+                                tracing::info!("shutting down repo");
+                                break;
+                            },
+                            Err(e) => {
+                                tracing::error!("Error receiving control event: {:?}", e);
                             }
                         }
                     },
                 }
             }
         })
     }
+
+    pub fn handle(&self) -> RepoHandle {
+        RepoHandle {
+            control_sender: self.control_sender.clone().expect("Repo has shut down."),
+            collection_sender: self.collection_sender.clone().expect("Repo has shut down."),
+            outboxes: self.outboxes.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct RepoHandle {
+    control_sender: Sender<ControlEvent>,
+    collection_sender: Sender<(CollectionId, CollectionEvent)>,
+    outboxes: Arc<Mutex<HashMap<CollectionId, Outbox>>>,
```

Comment:
I would rather do `Arc<Mutex<HashMap<RepoId, Outbox>>>`, and we may as well do Arc<Mutex<HashMap<RepoId, Arc<Mutex<Outbox>>>>>`